flink-libraries/flink-ml/src/test/scala/org/apache/flink/ml/recommendation/ImplicitALSTest.scala

```diff
@@ -0,0 +1,171 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.recommendation
+
+import org.apache.flink.ml.util.FlinkTestBase
+import org.scalatest._
+
+import scala.language.postfixOps
+import org.apache.flink.api.scala._
+import org.apache.flink.core.testutils.CommonTestUtils
+
+class ImplicitALSTest
+  extends FlatSpec
+    with Matchers
+    with FlinkTestBase {
+
+  override val parallelism = 2
+
+  behavior of "The modification of the alternating least squares (ALS) implementation" +
+    "for implicit feedback datasets."
+
+  it should "properly compute Y^T * Y, and factorize matrix" in {
+    import ExampleMatrix._
+
+    val rand = scala.util.Random
+    val numBlocks = 3
+    // randomly split matrix to blocks
+    val blocksY = Y
+      // add a random block id to every row
+      .map { row =>
+        (rand.nextInt(numBlocks), row)
+      }
+      // get the block via grouping
+      .groupBy(_._1).values
+      // add a block id (-1) to each block
+      .map(b => (-1, b.map(_._2)))
+      .toSeq
+
+    // use Flink to compute YtY
+    val env = ExecutionEnvironment.getExecutionEnvironment
+
+    val distribBlocksY = env.fromCollection(blocksY)
+
+    val YtY = ALS
+      .computeXtX(distribBlocksY, factors)
+      .collect().head
+
+    // check YtY size
+    YtY.length should be (factors * (factors - 1) / 2 + factors)
+
+    // check result is as expected
+    expectedUpperTriangleYtY
+      .zip(YtY)
+      .foreach { case (expected, result) =>
+        result should be (expected +- 0.1)
+      }
+
+    // temporary directory to avoid too few memory segments
+    val tempDir = CommonTestUtils.getTempDir + "/"
+
+    // factorize matrix with implicit ALS
+    val als = ALS()
+      .setIterations(iterations)
+      .setLambda(lambda)
+      .setBlocks(blocks)
+      .setNumFactors(factors)
+      .setImplicit(true)
+      .setAlpha(alpha)
+      .setSeed(seed)
+      .setTemporaryPath(tempDir)
+
+    val inputDS = env.fromCollection(implicitRatings)
+
+    als.fit(inputDS)
+
+    // check predictions on some user-item pairs
+    val testData = env.fromCollection(expectedResult.map{
+      case (userID, itemID, rating) => (userID, itemID)
+    })
+
+    val predictions = als.predict(testData).collect()
+
+    predictions.length should equal(expectedResult.length)
+
+    val resultMap = expectedResult map {
+      case (uID, iID, value) => (uID, iID) -> value
+    } toMap
+
+    predictions foreach {
+      case (uID, iID, value) => {
+        resultMap.isDefinedAt((uID, iID)) should be(true)
+
+        value should be(resultMap((uID, iID)) +- 1e-5)
+      }
+    }
+
+  }
+
+}
+
+object ExampleMatrix {
```

Comment:
Data should go to the `Recommendation.scala` file, as with the plain ALS matrix.